ThesesFR.js

```diff
@@ -79,24 +85,80 @@ function doWeb(doc, url) {
+	
+	// Each thesis record has an underlying .rdf file
+	Zotero.Utilities.HTTP.doGet(xmlDocumentUrl, function (text) {
+		const parser = new DOMParser();
+		const xmlDoc = parser.parseFromString(text, 'application/xml');
+
+		// Skiping invalid or empty RDF files : prevents crashes while importing multiple records
+		if (xmlDoc.getElementsByTagName('parsererror')[0] || xmlDoc.children[0].childElementCount === 0) {
+			Z.debug('Skiping: Invalid or empty RDF file.');
+			return;
 		}
-			
-		item.complete();
-	});
-	translator.getTranslatorObject(function (trans) {
-		trans.itemType = "thesis";
-		trans.doWeb(doc, url);
+		
+		// Importing XML namespaces for parsing purposes
+		const ns = {
+			bibo: 'http://purorg/ontology/bibo/',
+			dc: 'http://purl.org/dc/elements/1.1/',
+			dcterms: 'http://purl.org/dc/terms/',
+			foaf: 'http://xmlns.com/foaf/0.1/',
+			marcrel: 'http://www.loc.gov/loc.terms/relators/',
+			rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#'
+		};
+	
+		const title = ZU.xpathText(xmlDoc, '//dc:title', ns);
+		
+		// The case should't exists : preve;nts crashes while importing multiple records
+		if (!title) return;
+
+		const newItem = new Zotero.Item();
+		newItem.itemType = 'thesis';
+		newItem.title = title;
+
+		ZU.xpath(xmlDoc, '//marcrel:aut//foaf:Person/foaf:name | //marcrel:dis//foaf:Person/foaf:name', ns).forEach((auth) => {
+			const author = ZU.cleanAuthor(auth.textContent, 'author', true);
+			newItem.creators.push(author);
+		});
+
+		// Supervisor(s) must be considered as contributor(s) for french thesis
+		ZU.xpath(xmlDoc, '//marcrel:ths//foaf:Person/foaf:name', ns).forEach((sup) => {
+			const supervisor = ZU.cleanAuthor(sup.textContent, 'contributor', true);
+			newItem.creators.push(supervisor);
+		});
+
+		const abstractNotes = ZU.xpath(xmlDoc, '//dcterms:abstract', ns);
+		newItem.abstractNote = abstractNotes.length > 0 ? abstractNotes[0].textContent : undefined;
+
+		// '/s + digit' in url means thesis in preparation
+		newItem.thesisType = url.match(/\/s\d+/) ? 'These en préparation' : 'These de doctorat';
+
+		const organizations = ZU.xpath(xmlDoc, '//marcrel:dgg/foaf:Organization/foaf:name', ns);
+		newItem.university = organizations.length > 0 ? organizations[0].textContent : undefined;
+
+		const fullDate = ZU.xpathText(xmlDoc, '//dcterms:dateAccepted', ns);
+		const year = ZU.xpathText(xmlDoc, '//dc:date', ns);
+
+		// Some old records doesn't have a full date instead we can use the defense year
+		newItem.date = fullDate ? fullDate : year;
+
+		const permalink = ZU.xpath(xmlDoc, '//bibo:Document', ns);
+		newItem.url = permalink.length > 0 ? permalink[0].getAttribute('rdf:about') : url;
+		newItem.libraryCatalog = 'theses.fr';
+		newItem.rights = 'Les données de Theses.fr sont sous licence Etalab';
+
+		// Keep extra information such as laboratory, graduate schools, etc.
+		newItem.extra = Array.from(doc.getElementsByClassName('donnees-ombre')).map((description) => {
```

Comment:
This is better suited for a note than `extra`

Reply:
In this case I'll use a note instead